### PR TITLE
bugfix: set random seed for each worker process at `init_worker` phase, only `init` phase is not enough.

### PR DESCRIPTION
**Pull request by membphis**
### What this PR does / why we need it:

related PR: https://github.com/apache/apisix/pull/2306

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

**Comment by membphis** [comment]
@zlm0125 welcome to review and make a test with this PR. I think it should work fine for your case

**Comment by moonming** [comment]
why only `init` phase is not enough?

**Comment by membphis** [comment]
> why only `init` phase is not enough?

`init` works in `master` process if the [lua_code_cache](https://github.com/openresty/lua-nginx-module#lua_code_cache) is `on`.

We want to set different `seed` for different work processes, so we should use `init_worker` phase.

we can delete those code https://github.com/apache/apisix/pull/2357/files#diff-d982d52466e7c93c7b604358339b2a29R85-R90 . 
And run the test case, then we will get the same random number which is wrong.

image

Here is the right one:

image


image

**Comment by moonming** [comment]
> related PR: #2306

I don't understand what it has to do with this PR

**Comment by moonming** [comment]
And we should remove random in `init` phase if add random in `init_worker` phase.

**Comment by membphis** [comment]
> And we should remove random in `init` phase if add random in `init_worker` phase.

you are right. let me do this later.

File: apisix/init.lua [hunk 1]
```diff
@@ -82,6 +75,14 @@ end
 
 
 function _M.http_init_worker()
+    local seed, err = core.utils.get_seed_from_urandom()
+    if not seed then
+        core.log.warn('failed to get seed from urandom: ', err)
+        seed = ngx_now() * 1000 + ngx.worker.pid()
+    end
+    math.randomseed(seed)
+    core.log.info("random test in [1, 10000]: ", math.random(1, 1000000))
```
Comment (moonming) [inline apisix/init.lua:84]: should be removed

File: apisix/init.lua [hunk 2]
```diff
@@ -759,6 +760,14 @@ end
 
 function _M.stream_init_worker()
     core.log.info("enter stream_init_worker")
+    local seed, err = core.utils.get_seed_from_urandom()
+    if not seed then
+        core.log.warn('failed to get seed from urandom: ', err)
+        seed = ngx_now() * 1000 + ngx.worker.pid()
+    end
+    math.randomseed(seed)
+    core.log.info("random stream test in [1, 10000]: ", math.random(1, 1000000))
```
Comment (moonming) [inline apisix/init.lua:769]: ditto

File: apisix/init.lua (hunk 1 above)
Comment (membphis) [reply apisix/init.lua:84]: why do we need to remove this line?
It is necessary for debugging.

Comment (moonming) [reply apisix/init.lua:84]: Can it help user to find any problems?

Comment (membphis) [reply apisix/init.lua:84]: it not for the user, the developer need it for testing.

here is the code link of the test case: https://github.com/apache/apisix/pull/2357/files#diff-a46ca98b2569d841ad4742ef90427a1fR39

**Review by moonming — APPROVED** [review]

File: apisix/init.lua (hunk 1 above)
Comment (spacewander) [reply apisix/init.lua:84]: @membphis 
I think we should add a comment for it.

**Review by spacewander — APPROVED** [review]

**Pull request merged**